File: src/main/java/com/intgrah/algorithms/sort/HeapSort.java
package com.intgrah.algorithms.sort;

import java.util.Comparator;

import static com.intgrah.algorithms.util.Util.swap;

public class HeapSort<T> extends Sorter<T> {

    public HeapSort(Comparator<T> ord) {
        super(ord);
    }

    @Override
    public void sort(T[] a) {
        int start = a.length / 2;
        int end = a.length;
        while (end > 1) {
            if (start == 0) {
                end--;
                swap(a, 0, end);
            } else {
                start--;
            }
            int i = start;
            while (true) {
                int l = 2 * i + 1;
                int r = 2 * i + 2;
                if (r < end && ord.compare(a[l], a[r]) < 0 && ord.compare(a[i], a[r]) < 0) {
                    swap(a, i, r);
                    i = r;
                } else if (l < end && ord.compare(a[i], a[l]) < 0) {
                    swap(a, i, l);
                    i = l;
                } else {
                    break;
                }
            }
        }
    }

}

File: sort/heap_sort.py
from util import swap


def heap_sort(a: list):
    start = len(a) // 2 # First non-leaf node
    end = len(a)
    while end > 1:
        if start == 0:
            end -= 1
            swap(a, 0, end)
        else:
            start -= 1
        i = start # Sift down
        while True:
            l = 2 * i + 1
            r = 2 * i + 2
            if r < end and a[l] <= a[r] > a[i]:
                swap(a, i, r)
                i = r
            elif l < end and a[l] > a[i]:
                swap(a, i, l)
                i = l
            else:
                break
